File: main/path-sum-iii/path-sum-iii-fast.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, total):
        """
        :type root: TreeNode
        :type total: int
        :rtype: int
        """
        if root is None:
            return 0

        history = collections.Counter((0,)) # old subtotals
        cur = 0 # current subtotal
        count = 0 # the number of paths found that sum to the total

        def dfs(node):
            nonlocal cur, count

            cur += node.val
            count += history[cur - total] # count prevs where cur-prev==total
            history[cur] += 1

            if node.left is not None:
                dfs(node.left)
            if node.right is not None:
                dfs(node.right)

            history[cur] -= 1
            cur -= node.val

        dfs(root)
        return count

File: main/path-sum-iii/path-sum-iii-fast.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int PathSum(TreeNode root, int sum) {
        if (root == null) return 0;

        var history = new Dictionary<int, int> { { 0, 1 } }; // old subtotals
        var cur = 0; // current subtotal
        var count = 0; // the number of paths found that add to the desired sum

        void Dfs(TreeNode node) {
            cur += node.val;

            // Count prevs where cur - prev == total.
            if (history.TryGetValue(cur - sum, out var delta))
                count += delta;

            history.TryGetValue(cur, out var reps); // gives 0 if absent
            history[cur] = reps + 1;

            if (node.left != null) Dfs(node.left);
            if (node.right != null) Dfs(node.right);

            --history[cur];
            cur -= node.val;
        }

        Dfs(root);
        return count;
    }
}
